### PR TITLE
bws/98 + download csv file with wokrer credential

### PR DESCRIPTION
**Pull request by bohdanshcherba**

### CODE REVIEW

**Review by what1s1ove — CHANGES_REQUESTED** [review]
we need to use `saveAs` package and create a service for it

File: frontend/package.json [hunk 1]
```diff
@@ -13,12 +13,15 @@
     "build": "react-scripts build"
   },
   "dependencies": {
+    "-": "^0.0.1",
```
Comment (what1s1ove) [inline frontend/package.json]: ?

File: frontend/package-lock.json [hunk 2]
```diff
@@ -51,6 +54,11 @@
         "npm": ">=8 <9"
       }
     },
+    "node_modules/-": {
```
Comment (what1s1ove) [inline frontend/package-lock.json]: uninstall this packages

File: frontend/src/services/save-csv/save-csv.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { saveAs } from 'file-saver';
```
Comment (what1s1ove) [inline frontend/src/services/save-csv/save-csv.ts]: move to `frontend/src/services/saver/saver.service.ts`

File: frontend/src/store/eam-worker-configurate/actions.ts [hunk 4]
```diff
@@ -17,13 +17,21 @@ const workerCreate = createAsyncThunk<
...
       ...payload,
-      password: getRandomId(),
+      password,
     });
 
+    const content = [
+      ['name:', payload.name],
+      ['password:', password],
+    ];
+    saveCsv.save(content, 'Credentials.csv');
```
Comment (what1s1ove) [inline frontend/src/store/eam-worker-configurate/actions.ts]:
```suggestion
    saveCsv.save([
      ['name:', payload.name],
      ['password:', password],
    ], `${payload.name}-worker-credentials`);
```
also, add enums for the column names

File: frontend/src/services/services.ts [hunk 5]
```diff
@@ -20,6 +21,8 @@ const http = new Http({ storage });
 
 const notification = new Notification();
 
+const saveCsv = new SaveCsv();
```
Comment (what1s1ove) [inline frontend/src/services/services.ts]:
```suggestion
const saver = new Saver();
```

File: frontend/src/services/save-csv/save-csv.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+class SaveCsv {
+  public save(content: string[][], fileName: string): void {
+    const csvContent = content.map((e) => e.join(' ')).join('\n');
+
+    const file = new File([csvContent], fileName, {
+      type: 'text/csv;encoding:utf-8',
+    });
+    saveAs(file);
```
Comment (what1s1ove) [inline frontend/src/services/save-csv/save-csv.ts]:
```suggestion
    saveAs(new File([csvContent], `${fileName}.csv `, {
      type: ContentType.CSV,
    }));
```

File: frontend/src/services/save-csv/save-csv.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import { saveAs } from 'file-saver';
+
+class SaveCsv {
+  public save(content: string[][], fileName: string): void {
```
Comment (what1s1ove) [inline frontend/src/services/save-csv/save-csv.ts]:
```suggestion
  public saveCSV(content: string[][], fileName: string): void {
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]
image

we need to add icon button here. It should appear after worker create

File: frontend/src/services/save-csv/save-csv.ts (hunk 7 above)
Comment (bohdanshcherba) [reply frontend/src/services/save-csv/save-csv.ts]: svg?

Comment (what1s1ove) [reply frontend/src/services/save-csv/save-csv.ts]: svg? 🤔

**Comment by bohdanshcherba** [comment]
what should this button be?
image

File: frontend/package.json [hunk 8]
```diff
@@ -16,9 +16,11 @@
     "@hookform/error-message": "2.0.0",
     "@hookform/resolvers": "2.8.5",
     "@reduxjs/toolkit": "1.7.1",
+    "@types/file-saver": "^2.0.5",
```
Comment (what1s1ove) [inline frontend/package.json]: B19

File: frontend/package.json [hunk 9]
```diff
@@ -16,9 +16,11 @@
...
     "bws-shared": "file:../shared/build",
     "clsx": "1.1.1",
     "env-cmd": "10.1.0",
+    "file-saver": "^2.0.5",
```
Comment (what1s1ove) [inline frontend/package.json]: B19

File: frontend/src/common/enums/app/csv.enum.ts [hunk 10]
```diff
@@ -0,0 +1,6 @@
+enum CSV {
```
Comment (what1s1ove) [inline frontend/src/common/enums/app/csv.enum.ts]:
```suggestion
enum EAMCreateWorkerCSVColumn {
```
move to `frontend/src/common/enums/eam`

File: frontend/src/components/eam-create-worker/eam-create-worker.tsx [hunk 11]
```diff
@@ -1,33 +1,35 @@
...
-import { Input, Button, Table } from 'components/common/common';
+import { Button, Input, Table } from 'components/common/common';
 import { EAMWorkerCreateRequestDto } from 'common/types/types';
 import { EamWorkerCreate as CreateWorkerValidationSchema } from 'validation-schemas/validation-schemas';
 import styles from './styles.module.scss';
 import { EAMWorkerConfigurate as EAMWorkerConfigurateActions } from 'store/actions';
 import { DEFAULT_PAYLOAD } from './common/constants';
-import { getRows, getColumns } from './helpers/helpers';
+import { getColumns, getRows } from './helpers/helpers';
+import { saveAs } from 'file-saver';
```
Comment (what1s1ove) [inline frontend/src/components/eam-create-worker/eam-create-worker.tsx]:
```suggestion
```

Comment (what1s1ove) [reply frontend/src/components/eam-create-worker/eam-create-worker.tsx]: you should do it in the action

File: frontend/src/components/eam-create-worker/eam-create-worker.tsx [hunk 12]
```diff
@@ -111,6 +119,15 @@ const EAMWorkerCreate: FC = () => {
...
           </ul>
+          <div className={!csvFile ? 'visually-hidden' : ''}>
+            <button
+              className={styles.saveBtn}
+              type={ButtonType.BUTTON}
+              onClick={handleSave}
+            >
+              Save csv
+            </button>
+          </div>
```
Comment (what1s1ove) [inline frontend/src/components/eam-create-worker/eam-create-worker.tsx]:
```suggestion
const hasCsvColumns = Boolean(csvColumns.lenght)
          hasCsvColumns &&
            <button
              className={styles.saveBtn}
              type={ButtonType.BUTTON}
              onClick={handleSave}
            >
              Save csv
            </button>
```

File: frontend/src/store/eam-worker-configurate/actions.ts [hunk 13]
```diff
@@ -1,34 +1,41 @@
...
     });
 
-    navigation.push(AppRoute.EAM);
+    const csvFile = saver.saveCSV(
+      [
+        [CSV.NAME, payload.name],
+        [CSV.PASSWORD, password],
+      ],
+      `${payload.name}-worker-credentials`,
+    );
```
Comment (what1s1ove) [inline frontend/src/store/eam-worker-configurate/actions.ts]: on this step, you need to store columns in the storage, and then by clicking on the `download creds button`, you need to call another action that will trigger saving

File: frontend/src/store/eam-worker-configurate/reducer.ts [hunk 14]
```diff
@@ -6,16 +6,18 @@ import { EAMGroupGetByTenantResponseItemDto } from 'common/types/types';
 type State = {
   dataStatus: DataStatus;
   groups: EAMGroupGetByTenantResponseItemDto[];
+  csvFile: File | null;
```
Comment (what1s1ove) [inline frontend/src/store/eam-worker-configurate/reducer.ts]:
```suggestion
  csvColumns: string[][];
```

File: frontend/src/services/saver/saver.service.ts [hunk 15]
```diff
@@ -0,0 +1,13 @@
+import { ContentType } from 'common/enums/enums';
```
Comment (what1s1ove) [inline frontend/src/services/saver/saver.service.ts:1]: import of `saveAs` should be here

**Review by what1s1ove — CHANGES_REQUESTED** [review]
attach screenshot

File: frontend/src/store/eam-worker-configurate/actions.ts [hunk 16]
```diff
@@ -1,34 +1,50 @@
...
+
+const saveCSV = createAsyncThunk<void, void, AsyncThunkConfig>(
+  ActionType.SAVE_CSV,
+  async (_payload, { extra, getState }) => {
+    const { saver } = extra;
+    const { EAMWorkerConfigurate } = getState();
+
+    const { csvColumns } = EAMWorkerConfigurate;
+
+    saver.saveCSV(csvColumns, `${csvColumns[0][1]}-worker-credentials`);
```
Comment (what1s1ove) [inline frontend/src/store/eam-worker-configurate/actions.ts]:
```suggestion
   const [header: [name]] = csvColumns;

    saver.saveCSV(csvColumns, `${name}-worker-credentials`);
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Comment by what1s1ove** [comment]
> what should this button be? image

right, just align all button by a horisontal center

File: frontend/src/store/eam-worker-configurate/actions.ts (hunk 16 above)
Comment (bohdanshcherba) [reply frontend/src/store/eam-worker-configurate/actions.ts]: it not working

Comment (bohdanshcherba) [reply frontend/src/store/eam-worker-configurate/actions.ts]: maybe it?
```
const [[,name]] = csvColumns;
saver.saveCSV(csvColumns, `${name}-worker-credentials`);
```

Comment (what1s1ove) [reply frontend/src/store/eam-worker-configurate/actions.ts]: yes, you're right!

**Comment by bohdanshcherba** [comment]
image

File: frontend/src/components/eam-create-worker/eam-create-worker.tsx [hunk 17]
```diff
@@ -111,16 +118,29 @@ const EAMWorkerCreate: FC = () => {
...
           </ul>
-          <div className={styles.buttons}>
-            <div className={styles.button}>
-              <Button
-                btnStyle={ButtonStyle.OUTLINED}
-                label="Cancel"
-                to={AppRoute.EAM}
-              />
+          <div className={styles.buttonContainer}>
+            <div>
```
Comment (what1s1ove) [inline frontend/src/components/eam-create-worker/eam-create-worker.tsx]:
```suggestion
```
you don't need an addition wrapper

**Review by what1s1ove — APPROVED** [review]

File: frontend/src/components/eam-create-worker/eam-create-worker.tsx (hunk 17 above)
Comment (what1s1ove) [reply frontend/src/components/eam-create-worker/eam-create-worker.tsx]: remove it

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/components/eam-configurate-worker/styles.module.scss [hunk 18]
```diff
@@ -69,11 +69,27 @@
...
   gap: 10px;
 }
 
 .button {
   min-width: 200px;
 }
+
+.saveBtn {
+  position: absolute;
+  left: 0;
```
Comment (what1s1ove) [inline frontend/src/components/eam-configurate-worker/styles.module.scss]:
```suggestion
```

File: frontend/src/components/eam-configurate-worker/styles.module.scss [hunk 19]
```diff
@@ -69,11 +69,27 @@
 }
 
 .buttons {
+  position: relative;
```
Comment (what1s1ove) [inline frontend/src/components/eam-configurate-worker/styles.module.scss]:
```suggestion
```

File: frontend/src/components/eam-configurate-worker/eam-configurate-worker.tsx [hunk 20]
```diff
@@ -112,6 +119,15 @@ const EAMConfigurateWorker: FC = () => {
             </li>
           </ul>
           <div className={styles.buttons}>
+            {hasCsvColumns && (
```
Comment (what1s1ove) [inline frontend/src/components/eam-configurate-worker/eam-configurate-worker.tsx:122]:
```suggestion
<div className={styles.buttonContainer}>
            {hasCsvColumns && (
```
you need one, not two wrappers

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Pull request merged**